src/cocotb/_scheduler.py

```diff
@@ -172,61 +173,94 @@ def thread_wait(self):
-    schedule them in turn.
-
-    .. attention::
-
-       Implementors should not depend on the scheduling order!
+class SimPhase(Enum):
+    NORMAL = auto()
+    READ_WRITE = auto()
+    READ_ONLY = auto()
 
-    Due to the simulator nuances and fun with delta delays we have the
-    following modes:
 
-    Normal mode
-        - Callbacks cause tasks to be scheduled.
-        - Any pending writes are cached and do not happen immediately.
-
-    ReadOnly mode
-        - Corresponds to ``cbReadOnlySynch`` (VPI) or ``vhpiCbRepEndOfTimeStep``
-          (VHPI).  In this state we are not allowed to perform writes.
-
-    Write mode
-        - Corresponds to ``cbReadWriteSynch`` (VPI) or ``vhpiCbRepLastKnownDeltaCycle`` (VHPI)
-          In this mode we play back all the cached write updates.
-
-    We can legally transition from Normal to Write by registering a :class:`~cocotb.triggers.ReadWrite`
-    callback, however usually once a simulator has entered the ReadOnly phase
-    of a given timestep then we must move to a new timestep before performing
-    any writes.  The mechanism for moving to a new timestep may not be
-    consistent across simulators and therefore we provide an abstraction to
-    assist with compatibility.
-
-
-    Unless a task has explicitly requested to be scheduled in ReadOnly
-    mode (for example wanting to sample the finally settled value after all
-    delta delays) then it can reasonably be expected to be scheduled during
-    "normal mode" i.e. where writes are permitted.
+class Scheduler:
+    """The main Task scheduler.
+
+    How It Generally Works:
+        Tasks are `queued` to run in the scheduler with :meth:`_queue`.
+        Queueing adds the Task and an Outcome value to :attr:`_pending_tasks`.
+        The main scheduling loop is located in :meth:`_event_loop` and loops over the queued Tasks and `schedules` them.
+        :meth:`_schedule` schedules a Task -
+        continuing its execution from where it previously yielded control -
+        by injecting the Outcome value associated with the Task from the queue.
+        The Task's body will run until it finishes or reaches the next ``await`` statement.
+        If a Task reaches an ``await``, :meth:`_schedule` will convert the value yielded from the Task into a Trigger with :meth:`_trigger_from_any` and its friend methods.
+        Triggers are then `primed` (with :meth:`~cocotb.triggers.Trigger._prime`)
+        with a `react` function (:meth:`_sim_react` or :meth:`_react)
+        so as to wake up Tasks waiting for that Trigger to `fire` (when the event encoded by the Trigger occurs).
+        This is accomplished by :meth:`_resume_task_upon`.
+        :meth:`_resume_task_upon` also associates the Trigger with the Task waiting on it to fire by adding them to the :attr:`_triggers2tasks` map.
+        If instead, a Task finishes, :meth:`_schedule` will cause the :class:`~cocotb.triggers.Join` trigger to fire.
```

Comment:
```suggestion
        If, instead of reaching an ``await``, a Task finishes, :meth:`_schedule` will cause the :class:`~cocotb.triggers.Join` trigger to fire.
```